src/Microsoft.JSInterop/DotNetDispatcher.cs

```diff
@@ -0,0 +1,181 @@
+        /// <param name="methodIdentifier">The identifier of the method to be invoked. The method must be annotated with a <see cref="JSInvokableAttribute"/> matching this identifier string.</param>
+        /// <param name="argsJson">A JSON representation of the parameters.</param>
+        /// <returns>A JSON representation of the return value, or null.</returns>
+        public static string Invoke(string assemblyName, string methodIdentifier, string argsJson)
+        {
+            var syncResult = InvokeSynchronously(assemblyName, methodIdentifier, argsJson);
+            return syncResult == null ? null : Json.Serialize(syncResult);
+        }
+
+        /// <summary>
+        /// Receives a call from JS to .NET, locating and invoking the specified method asynchronously.
+        /// </summary>
+        /// <param name="callId">A value identifying the asynchronous call that should be passed back with the result, or null if no result notification is required.</param>
+        /// <param name="assemblyName">The assembly containing the method to be invoked.</param>
+        /// <param name="methodIdentifier">The identifier of the method to be invoked. The method must be annotated with a <see cref="JSInvokableAttribute"/> matching this identifier string.</param>
+        /// <param name="argsJson">A JSON representation of the parameters.</param>
+        /// <returns>A JSON representation of the return value, or null.</returns>
+        public static void BeginInvoke(string callId, string assemblyName, string methodIdentifier, string argsJson)
+        {
+            var syncResult = InvokeSynchronously(assemblyName, methodIdentifier, argsJson);
+
+            // If there was no callId, the caller does not want to be notified about the result
+            if (callId != null)
+            {
+                // Invoke and coerce the result to a Task so the caller can use the same async API
+                // for both synchronous and asynchronous methods
+                var task = syncResult is Task syncResultTask ? syncResultTask : Task.FromResult(syncResult);
+                task.ContinueWith(completedTask =>
+                {
+                    // DotNetDispatcher only works with JSRuntimeBase instances.
+                    // If the developer wants to use a totally custom IJSRuntime, then their JS-side
+                    // code has to implement its own way of returning async results.
+                    var jsRuntimeBaseInstance = (JSRuntimeBase)JSRuntime.Current;
+
+                    try
+                    {
+                        var result = TaskGenericsUtil.GetTaskResult(completedTask);
+                        jsRuntimeBaseInstance.EndInvokeDotNet(callId, true, result);
+                    }
+                    catch (Exception ex)
+                    {
+                        ex = UnwrapException(ex);
+                        jsRuntimeBaseInstance.EndInvokeDotNet(callId, false, ex.ToString());
```

Comment:
Is there any use case here for passing the exception object? It seems little odd to me that this layer would decide how you want to treat exceptions.

Reply:
That's a good point. Agreed and updated.